### ISSUE TITLE
Standardize API

### ISSUE THREAD

**Issue by lnproxy** [niteshbalusu11/lnproxy-ts#2]
Would be good to have a standard API for all implementations so that integrations like robosats' (@Reckless-Satoshi) don't have to have different code for relays running the original Go implementation or this new typescript implementation.

Do you think it would make sense to collaborate on a simple spec of some kind?

**Comment by niteshbalusu11** [comment]
Sure, I'm up for that.

**Comment by lnproxy** [comment]
Great!

I think the ideal would be a simple repo hosting a document like [this](https://github.com/lnurl/luds/blob/luds/06.md) describing the flow of requesting a wrapped invoice in plain and precise language.

One thing we'd have to decide is what the request looks like.  We have two options:
```
curl https://lnproxy.org/api/ADD_BOLT_11_INVOICE_HERE
```
with additional parameters as query parameters, and
```
curl --header "Content-Type: application/json" \
  --request POST \
  --data '{"request":"ADD_BOLT_11_INVOICE_HERE"}' \
  http://localhost:4545
```
which could support additional parameters in the data object.  I have a slight preference for the former, more minimal interface, but can appreciate that it could be awkward for long invoices (e.g. with lot's of route hints), specially if we need to pass additional parameters for setting a routing budget or, for example, changing the invoice's description.

Another thing to decide on would be additional features like the ability to set a custom routing budget (the additional amount of satoshis that will be used by the relay to route the payment).  But maybe these can just have their own sub documents like in the [lnurl spec](https://github.com/lnurl/luds).

Please let me know what you would prefer.  Then I can write something up and ask you to review it once it's ready.

**Comment by niteshbalusu11** [comment]
If we are taking the lnurl route then I think we have to use GET request with url encoded query params because I believe all of lnurl uses that. I personally prefer POST with data as object in body over URLs with query params however I understand the simplicity of URLs. 
You can take a call on that, whatever you prefer. 

The way I designed mine with respect to fees is, I haven't added an option for users to add their own max_fee or max_fee_rate because I probe the route before giving out the hodl invoice and add the fees needed to pay the original invoice and add a 2sat premium for running the service. 
Of course this comes with the assumption that the same route still has liquidity available when actually settling the invoices. 

cltv_delta: we need a high value for this, I am adding the delta of all hops and and then adding some extra buffer on top of that. 

What else needs to go in the spec in your opinion?

**Comment by lnproxy** [comment]
> If we are taking the lnurl route then I think we have to use GET request with url encoded query params because I believe all of lnurl uses that. I personally prefer POST with data as object in body over URLs with query params however I understand the simplicity of URLs. You can take a call on that, whatever you prefer.

OK, sounds good.
 
> The way I designed mine with respect to fees is, I haven't added an option for users to add their own max_fee or max_fee_rate because I probe the route before giving out the hodl invoice and add the fees needed to pay the original invoice and add a 2sat premium for running the service. Of course this comes with the assumption that the same route still has liquidity available when actually settling the invoices.
> 
> cltv_delta: we need a high value for this, I am adding the delta of all hops and and then adding some extra buffer on top of that.

That's really cool.  Much better UX and less failed payments.  Probably the biggest benefit of probing is in setting the `cltv_delta`.  My implementation just asks for a large one and tells lnd not to exceed it when paying the original invoice (delta some buffer). I think I'll implement probing too.

> What else needs to go in the spec in your opinion?

I think it would be nice to have an option to allow users to change the invoice description.  If you're using a custodial wallet, and don't want the custodian to know you're paying some service, you can use an lnproxy relay to obscure that but if the service puts their name in the invoice description then you're out of luck.  At the same time, you don't want to always produce blank invoice descriptions since you need specific ones for things like lnurl.

I don't have anything else in mind besides that.

**Comment by niteshbalusu11** [comment]
Cool, yeah setting a description sounds good. We could draft a V1 of the spec and then take it from there.

**Comment by lnproxy** [comment]
Spec is almost ready for your review but I'm stuck on one small detail.

One of the nice use cases of an lnproxy relay is the ability to set up a personal lightning address server using one of your domains without revealing the pubkey of your node:
    https://github.com/lnproxy/lnproxy-address
    https://github.com/nbd-wtf/satdress/pull/42

In order for this to work, the lnproxy relay needs to be able to return invoices of an amount known in advance. (The reason for this is that users pick the amount in the lnurl protocol and wallets verify that the invoices from lnurl servers have the specified amount).  This is possible when the relay allows you to set a custom routing budget (you just request an invoice amount and routing budget that add up to the lnurl amount).

One way to do it would be to request an invoice to get a fee estimate from the probe, then request another invoice with a smaller amount and hope the probe comes up with the same fee.  It would probably work well since I think most implementations will remember successful routes.

Any thoughts?

**Comment by niteshbalusu11** [comment]
We could still add custom routing budgets. I don't think probing should be part of the spec, we can't really control how people want to do their pathfinding and other stuff. 
I went the probing way because I felt the invoice amount can be more accurate.

**Comment by lnproxy** [comment]
Here's my first draft:  https://github.com/lnproxy/spec/blob/main/README.md

**Comment by niteshbalusu11** [comment]
> User verifies that the amount in the proxy invoice is exactly equal to the sum of the amount in the original invoice and the amount in the routing_msat request.

This part is tricky because, what if the proxy service wants to charge a fee, which in most cases you would want to because I'm not sure you wanna run a web server for free.

**Comment by lnproxy** [comment]
The routing_msat parameter is just the maximum amount the relay can use to route.  It can pick a cheaper route and take a fee or if routing_msat is too low, just fail the payment.  The only necessary part is that a user is able to get a proxy invoice of some specific amount.

**Comment by niteshbalusu11** [comment]
The user verifying the sum to be exactly equal to the routing_msat and original invoice amount could be misleading imo, maybe also specify and any additional fee the proxy node can charge in brackets or something?

**Comment by lnproxy** [comment]
Yes, I see what you mean.

I think there's lots of reasons to think of the relay's fee as just another routing fee.  The obvious reason is technical correctness: the relay really is part of the payment route:

    payer (paying proxy invoice) -> -> -> relay (paying original invoice) -> -> -> destination

but due to the way routing fees work, it doesn't get a routing fee unless it adds one amount in the proxy invoice (or sets up some kind of virtual node to collect a virtual routing fee).

I added a sentence to clarify that the payment will fail if `routing_msat` is too low to cover the "costs of routing the payment" (which, I would argue, includes any fee charged by the relay):
https://github.com/lnproxy/spec/commit/9e4df9d643e5015bb2bf5666ac0f69ca25e2acb4 

This may still be too vague.  If you have specific language in mind, I'm happy to use whatever you think is best.

**Comment by niteshbalusu11** [comment]
I don’t think that terminology is wrong, what a proxy node can charge is basically any number you put on the invoice. An arbitrary number could be added or a percentage of the total invoice amount could just be added to the invoice as a fee. I just feel it’s better to add that to not give the impression of the service being always free or something.

**Comment by lnproxy** [comment]
OK, I agree.
